<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Htrans : Horizontal Gene Transfer project">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Htrans</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/kheath/htrans">View on GitHub</a>

          <h1 id="project_title">Htrans</h1>
          <h2 id="project_tagline">Horizontal Gene Transfer project</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/kheath/htrans/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/kheath/htrans/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h1>
<a id="horizontal-gene-transfer-project" class="anchor" href="#horizontal-gene-transfer-project" aria-hidden="true"><span class="octicon octicon-link"></span></a>Horizontal Gene Transfer project</h1>

<h2>
<a id="collaborative-project-for-eliot-bush" class="anchor" href="#collaborative-project-for-eliot-bush" aria-hidden="true"><span class="octicon octicon-link"></span></a>Collaborative project for Eliot Bush</h2>

<h2>
<a id="goals" class="anchor" href="#goals" aria-hidden="true"><span class="octicon octicon-link"></span></a>Goals:</h2>

<ul>
<li>Partially reconstruct gene family lineages in Escherichia Coli</li>
<li>Identify horizontal gene transfer events</li>
</ul>

<h3>
<a id="superalignv2py--phylibalignpy" class="anchor" href="#superalignv2py--phylibalignpy" aria-hidden="true"><span class="octicon octicon-link"></span></a>superalignv2.py &amp; phylibalign.py</h3>

<p>These two python scripts are very similar.  They take a file with blocks of genes (.afa) and convert it to fasta format.</p>

<h3>
<a id="fasta2phylipsh" class="anchor" href="#fasta2phylipsh" aria-hidden="true"><span class="octicon octicon-link"></span></a>fasta2phylip.sh</h3>

<p>This is a shell script that converts fasta files to phylip format.</p>

<h3>
<a id="mrca" class="anchor" href="#mrca" aria-hidden="true"><span class="octicon octicon-link"></span></a>mrca</h3>

<p>mrca(tree, family) returns the most recent common ancestor for a given gene family.</p>

<h3>
<a id="testatree" class="anchor" href="#testatree" aria-hidden="true"><span class="octicon octicon-link"></span></a>testATree</h3>

<p>The correct phylogenetic tree for our 5 sample species.  I typed this out by hand based on a tree generated by RAxML, there was no script directly involved in making this.</p>

<h3>
<a id="processfamgenes" class="anchor" href="#processfamgenes" aria-hidden="true"><span class="octicon octicon-link"></span></a>processFamGenes</h3>

<p>This is a collection of scripts put together to generate the needed information
to run dupDel.  It's a good idea to run this once and keep the files around.
It is important to note that the input files are made once with other scripts I
haven't looked into.</p>

<p>Input:</p>

<ul>
<li>fam.out (silix results)</li>
<li>geneSpeciesMap.txt</li>
<li>dbList.txt (file with sample species)</li>
<li>geneOrder.txt </li>
</ul>

<p>Output:</p>

<ul>
<li>famGenes.txt</li>
<li>famInfoResult.txt</li>
<li>adjacencyInfo.txt</li>
</ul>

<p>Sample command: python processFamGenes.py -f fam.out -m geneSpeciesMap.txt -d dbList.txt -g geneOrder.txt</p>

<h3>
<a id="dupdel" class="anchor" href="#dupdel" aria-hidden="true"><span class="octicon octicon-link"></span></a>dupDel</h3>

<p>This script calculates the minimum cost and associated duplications/deletions for every gene family.</p>

<p>Input:</p>

<ul>
<li>testATree</li>
<li>famInfoResult.txt</li>
</ul>

<p>Output:</p>

<ul>
<li>dupDelAll.txt</li>
</ul>

<p>Sample command: python dupDel.py -t testATree -f famInfoResult.txt -d 3 -c 5 -n 1</p>

<h3>
<a id="htranspy" class="anchor" href="#htranspy" aria-hidden="true"><span class="octicon octicon-link"></span></a>htrans.py</h3>

<p>This is a wrapper for the main pipeline.  Make sure you've done all the preprocessing steps before running this (Preprocessing not included in this repository).</p>

<p>Example: python htrans.py -f siLiX_families -m gene&lt;-&gt;species_map -d list_of_species_of_interest -g gene_order -t phylogenetic_tree -b deletion_cost -c duplication_cost -s #_of_species -o full_species_list</p>

<h3>
<a id="authors--contributors" class="anchor" href="#authors--contributors" aria-hidden="true"><span class="octicon octicon-link"></span></a>Authors &amp; Contributors</h3>

<p>Kevin Heath &amp; Zunyan Wang</p>

<h3>
<a id="support-or-contact" class="anchor" href="#support-or-contact" aria-hidden="true"><span class="octicon octicon-link"></span></a>Support or Contact</h3>

<p>Email me at <a href="mailto:kevin.n.heath@gmail.com">kevin.n.heath@gmail.com</a></p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Htrans maintained by <a href="https://github.com/kheath">kheath</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
